File: pixi-build-backends/backends/pixi-build-ros/tests/integration/data/ros-workspace/src/navigator_implicit/src/navigator.cpp
#define _USE_MATH_DEFINES

#include <cmath>

#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/point.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "turtlesim/msg/pose.hpp"

class TurtleNavigator : public rclcpp::Node
{
public:
    TurtleNavigator()
        : Node("turtle_navigator"), x_goal_(4.0), y_goal_(5.0), kp_(1.0), ki_(0.0), kd_(0.05), prev_error_(0.0), integral_(0.0)
    {
        subscription_ = this->create_subscription<geometry_msgs::msg::Point>(
            "coordinates", 10, std::bind(&TurtleNavigator::goal_callback, this, std::placeholders::_1));
        pose_subscription_ = this->create_subscription<turtlesim::msg::Pose>(
            "turtle1/pose", 10, std::bind(&TurtleNavigator::pose_callback, this, std::placeholders::_1));
        publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("turtle1/cmd_vel", 10);

        timer_ = this->create_wall_timer(
            std::chrono::milliseconds(100), std::bind(&TurtleNavigator::control_loop, this));

        RCLCPP_INFO(this->get_logger(), "Turtle Navigator has been started!");
        RCLCPP_INFO(this->get_logger(), "Initial goal: x=%f, y=%f", x_goal_, y_goal_);
    }

private:
    void goal_callback(const geometry_msgs::msg::Point::SharedPtr msg)
    {
        x_goal_ = msg->x;
        y_goal_ = msg->y;
        RCLCPP_INFO(this->get_logger(), "Received goal: x=%f, y=%f", x_goal_, y_goal_);
    }

    void pose_callback(const turtlesim::msg::Pose::SharedPtr msg)
    {
        x_current_ = msg->x;
        y_current_ = msg->y;
        theta_current_ = msg->theta;
    }

    void control_loop()
    {
//        RCLCPP_INFO(this->get_logger(), "Hello Ruben!");
        double error_x = x_goal_ - x_current_;
        double error_y = y_goal_ - y_current_;
        double distance_error = std::sqrt(error_x * error_x + error_y * error_y);

        double angle_to_goal = std::atan2(error_y, error_x);
        double angle_error = angle_to_goal - theta_current_;

        // Normalize angle error to the range [-pi, pi]
        while (angle_error > M_PI) angle_error -= 2 * M_PI;
        while (angle_error < -M_PI) angle_error += 2 * M_PI;

        // PID control
        double control_signal = kp_ * distance_error + ki_ * integral_ + kd_ * (distance_error - prev_error_);
        integral_ += distance_error;
        prev_error_ = distance_error;

        // Limit control signal
        double max_linear_speed = 2.0; // Max linear speed
        double max_angular_speed = 2.0; // Max angular speed
        control_signal = std::clamp(control_signal, -max_linear_speed, max_linear_speed);

        // Publish velocity commands
        auto msg = geometry_msgs::msg::Twist();
        msg.linear.x = control_signal;
        msg.angular.z = 4.0 * angle_error; // simple P controller for angle
        msg.angular.z = std::clamp(msg.angular.z, -max_angular_speed, max_angular_speed);

        publisher_->publish(msg);
    }

    rclcpp::Subscription<geometry_msgs::msg::Point>::SharedPtr subscription_;
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr pose_subscription_;
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;

    double x_goal_, y_goal_;
    double x_current_, y_current_, theta_current_;
    double kp_, ki_, kd_;
    double prev_error_, integral_;
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<TurtleNavigator>());
    rclcpp::shutdown();
    return 0;
}

File: tests/data/pixi-build/ros-workspace/src/navigator_py/navigator_py/navigator.py
import math
from typing import Any

import rclpy  # type: ignore[import-not-found]
from geometry_msgs.msg import Point, Twist  # type: ignore[import-not-found]
from rclpy.node import Node  # type: ignore[import-not-found]
from turtlesim.msg import Pose  # type: ignore[import-not-found]


class TurtleNavigator(Node):  # type: ignore[misc]
    def __init__(self) -> None:
        super().__init__(node_name="turtle_navigator")
        self.x_goal = 5.0
        self.y_goal = 5.0
        self.kp = 1.0
        self.ki = 0.0
        self.kd = 0.05
        self.prev_error = 0.0
        self.integral = 0.0

        self.subscription = self.create_subscription(Point, "coordinates", self.goal_callback, 10)
        self.pose_subscription = self.create_subscription(
            Pose, "turtle1/pose", self.pose_callback, 10
        )
        self.publisher = self.create_publisher(Twist, "turtle1/cmd_vel", 10)

        self.timer = self.create_timer(0.1, self.control_loop)

        self.x_current = 0.0
        self.y_current = 0.0
        self.theta_current = 0.0

    def goal_callback(self, msg: Any) -> None:
        self.x_goal = msg.x
        self.y_goal = msg.y
        self.get_logger().info(f"Received goal: x={self.x_goal}, y={self.y_goal}")

    def pose_callback(self, msg: Any) -> None:
        self.x_current = msg.x
        self.y_current = msg.y
        self.theta_current = msg.theta

    def control_loop(self) -> None:
        error_x = self.x_goal - self.x_current
        error_y = self.y_goal - self.y_current
        distance_error = math.sqrt(error_x**2 + error_y**2)

        angle_to_goal = math.atan2(error_y, error_x)
        angle_error = angle_to_goal - self.theta_current

        # Normalize angle error to the range [-pi, pi]
        while angle_error > math.pi:
            angle_error -= 2 * math.pi
        while angle_error < -math.pi:
            angle_error += 2 * math.pi

        # PID control
        control_signal = (
            self.kp * distance_error
            + self.ki * self.integral
            + self.kd * (distance_error - self.prev_error)
        )
        self.integral += distance_error
        self.prev_error = distance_error

        # Limit control signal
        max_linear_speed = 2.0  # Max linear speed
        max_angular_speed = 2.0  # Max angular speed
        control_signal = max(min(control_signal, max_linear_speed), -max_linear_speed)

        # Publish velocity commands
        msg = Twist()
        msg.linear.x = control_signal
        msg.angular.z = 4.0 * angle_error  # Simple P controller for angle
        msg.angular.z = max(min(msg.angular.z, max_angular_speed), -max_angular_speed)

        self.publisher.publish(msg)


def main(args: Any = None) -> None:
    rclpy.init(args=args)
    print("Turtle Navigator")
    turtle_navigator = TurtleNavigator()
    print("Waiting for command..")
    rclpy.spin(turtle_navigator)
    turtle_navigator.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()
